### PR TITLE
Basic behavior

### PR DESCRIPTION
**Pull request by tcollier**
@kristjan @heidi @quarterdome 

NOTE: This is a public repository!

Take a CSV file and enqueue a single job to process it row-by-row. When the processing is complete, it invokes a callback on a handler class/module. The client is responsible for defining the handler and the item processor. See the <a href="https://github.com/apartmentlist/bulk-processor/tree/basic-behavior#bulkprocessor">README</a> for detailed usage info.

### CODE REVIEW

**Comment by quarterdome** [comment]
Nice, looks really solid.

File: README.md [hunk 1]
```diff
@@ -24,7 +24,103 @@ Or install it yourself as:
...
+  # @param successes [Hash<Fixnum, Array<String>>] keys are all successfully
+  #   processed rows, indexed from 0 (row 1 in the CSV is index 0 in this hash)
+  #   The values are arrays of messages the item processor generated for the row
+  #   (may be empty), e.g. { 0 => [], 1 => ['pet ID = 22 created'] }
+  # @param errors [Hash<Fixnum, Array<String>>] similar structure to successes,
+  #   but rows that were not completed successfully.
+  # @param fatal_error [StandardError] if nil, then all rows were processed,
+  #   else the error that was raise is passed in here
+  def self.complete(payload, successes, errors, fatal_error = nil)
+    if fatal_error
```
Comment (kristjan) [inline README.md]: Shouldn't `fatal_error` get communicated to someone?

File: lib/bulk_processor.rb [hunk 2]
```diff
@@ -1,5 +1,62 @@
...
+
+    if validator.valid?
+      records = table.map(&:to_hash)
+      Job.perform_later(records, item_processor.to_s, handler.to_s, payload)
+    else
+      @errors = validator.errors
+    end
+  rescue NoMethodError => error
+    if error.message == BAD_HEADERS_ERROR_MSG
+      @errors << 'Missing or malformed column header'
```
Comment (kristjan) [inline lib/bulk_processor.rb]: It'd be helpful to communicate the list of required and optional headers so the user can check their file

File: lib/bulk_processor/job.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class BulkProcessor
+  class Job < ActiveJob::Base
+    queue_as 'bulk_processor'
+
+    def perform(records, item_proccessor, handler, payload)
+      item_proccessor_class = item_proccessor.constantize
+      handler_class = handler.constantize
+
+      successes = {}
+      failures = {}
```
Comment (kristjan) [inline lib/bulk_processor/job.rb]: I suspect the endgame of these is a `BulkProcessor::Row` that knows its index, data, messages, etc.

File: spec/bulk_processor_spec.rb [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+describe BulkProcessor do
+  describe '.process usage' do
+    let(:required_columns) { [] }
+    let(:optional_columns) { ['name'] }
+    let(:stream) { StringIO.new(csv) }
+    let(:csv) { "name\nRex" }
+
+    before do
+      TestItemProcessor.required_columns = required_columns
+      TestItemProcessor.optional_columns = optional_columns
```
Comment (kristjan) [inline spec/bulk_processor_spec.rb]: This sets global state on the class for the whole test suite. It'd be safer to create a subclass of `TestItemProcessor` right here with methods that return what you want.

**Comment by kristjan** [comment]
LGTM. Only semi-blocker is global state in specs.

**Pull request merged**